qiskit/pulse/frame.py

```diff
@@ -0,0 +1,77 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Implements a Frame."""
+
+from typing import Optional, Tuple
+
+from qiskit.circuit import Parameter
+from qiskit.pulse.utils import validate_index
+from qiskit.pulse.exceptions import PulseError
+
+
+class Frame:
+    """A frame is a frequency and a phase."""
+
+    def __init__(self, identifier: str, parametric_index: Optional[Parameter] = None):
+        """
+        Args:
+            identifier: The index of the frame.
+            parametric_index: An optional parameter to specify the numeric part of the index.
+
+        Raises:
+            PulseError: if the frame identifier is not a string.
+        """
+        if parametric_index is not None:
+            validate_index(parametric_index)
+
+        if not isinstance(identifier, str):
+            raise PulseError(f"Frame identifiers must be string. Got {type(identifier)}.")
+
+        self._identifier = (identifier, parametric_index)
+        self._hash = hash((type(self), self._identifier))
+
+    @property
+    def identifier(self) -> Tuple:
+        """Return the index of this frame. The index is a label for a frame."""
+        return self._identifier
+
+    @property
+    def prefix(self) -> str:
+        """Return the prefix of the frame."""
+        return self._identifier[0]
```

Comment:
Is this required? Seems a bit odd?.

Reply:
With ff44ff8 this is required as the pulse parameter manager can handle frames by doing
```python
    def visit_Frame(self, node: Frame):
        """Assign parameters to ``Frame`` object."""
        if isinstance(node.identifier[1], ParameterExpression):
            new_index = self._assign_parameter_expression(node.identifier[1])

            if not isinstance(new_index, ParameterExpression):
                validate_index(new_index)

                return node.__class__(node.prefix, new_index)

        return node
```